test/framework/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java

```diff
@@ -400,10 +408,27 @@ protected NumberSyntheticSourceSupport(Function<Number, Number> round, boolean i
             this.ignoreMalformed = ignoreMalformed;
         }
 
+        @Override
+        public boolean preservesExactSource() {
+            // We opt in into fallback synthetic source if there is no doc values
+            // which preserves exact source.
+            return docValues == false;
+        }
+
         @Override
         public SyntheticSourceExample example(int maxVals) {
             if (randomBoolean()) {
                 Tuple<Object, Object> v = generateValue();
+                if (preservesExactSource()) {
+                    var rawInput = v.v1();
+
+                    // Confusing as it is this code actually runs
```

Comment:
Nit: I'd skip the "Confusing as it is" part, it's all confusing.. Same below.

Reply:
Writing that made me feel better but i'll remove it.